network/simple-onchain-discovery/src/builder.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::ConfigurationChangeListener;
+use channel::libra_channel;
+use libra_config::config::RoleType;
+use libra_types::on_chain_config::OnChainConfigPayload;
+use network::connectivity_manager::ConnectivityRequest;
+use tokio::runtime::Handle;
+
+struct ConfigurationChangeListenerConfig {
+    role: RoleType,
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    reconfig_events: libra_channel::Receiver<(), OnChainConfigPayload>,
+}
+
+impl ConfigurationChangeListenerConfig {
+    fn new(
+        role: RoleType,
+        conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+        reconfig_events: libra_channel::Receiver<(), OnChainConfigPayload>,
+    ) -> Self {
+        Self {
+            role,
+            conn_mgr_reqs_tx,
+            reconfig_events,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, PartialOrd)]
+enum State {
+    CREATED,
+    BUILT,
+    STARTED,
+}
+
+pub struct ConfigurationChangeListenerBuilder {
+    config: Option<ConfigurationChangeListenerConfig>,
+    listener: Option<ConfigurationChangeListener>,
+    state: State,
+}
+
+impl ConfigurationChangeListenerBuilder {
+    pub fn create(
+        role: RoleType,
+        conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+        reconfig_events: libra_channel::Receiver<(), OnChainConfigPayload>,
+    ) -> ConfigurationChangeListenerBuilder {
+        Self {
+            config: Some(ConfigurationChangeListenerConfig::new(
+                role,
+                conn_mgr_reqs_tx,
+                reconfig_events,
+            )),
+            listener: None,
+            state: State::CREATED,
+        }
+    }
+
+    pub fn build(&mut self) -> &mut Self {
+        assert_eq!(self.state, State::CREATED);
+        self.state = State::BUILT;
+        let config = self.config.take().expect("Listener must be configured");
+        self.listener = Some(ConfigurationChangeListener::new(
+            config.conn_mgr_reqs_tx,
+            config.reconfig_events,
+            config.role,
+        ));
+        self
+    }
```

Comment:
Is there a reason these can't be crushed into one?  Nothing is being configured between the two steps.

Reply:
going back to #4626, the end state will have main_node doing something similar to:

```
network.create()
consensus.create()
statesync.create();

....

network.build()
statesnc.build()
consensus.build()

...

consensus.start()
statesync.start()
network.start()
```

The key power is in decoupling the configure/build/start phases so that dependencies are easy to address in isolation.